Ruby translation of GnuPG-checking Git hook

creideiki opened the pull request:
The Git server-side hook checking GnuPG signatures is currently (see https://github.com/saab-simc-admin/scripts) written in Bash, which makes me nervous. This is a straight translation into Ruby, using Rugged and GPGME.

This needs a lot more testing before it can be accepted.

creideiki committed:
Ruby translation of GnuPG-checking Git hook

The Git server-side hook checking GnuPG signatures is currently (see
https://github.com/saab-simc-admin/scripts) written in Bash, which
makes me nervous. This is a straight translation into Ruby, using
Rugged and GPGME.

This needs a lot more testing before it can be accepted.

creideiki committed:
Move allowed signer list to separate file

Having the list of allowed signers hard-coded in the script is ugly
and hard to change. Move it to a separate file instead.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+        allowed = is_allowed_signer(signer)
+      end
+    rescue Exception => e
+      puts e
```
Is it really safe to continue if there is an exception?

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+      when :tag
+        # The ref is an annotated tag
+        if rev_old.to_i(16) == 0 and not repo.config['hooks.allowunsignedtags']
+          puts "*** Tag #{ref} already exists."
```
The doc at top says that the config options for this is `hooks.allowmodifytag`

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+            if allowed
+              puts "*** Good signature on tag #{ref} by #{signer}."
+            else
+              puts "*** Rejecting tag #{ref} due to lack of a valid GPG signature."
```
Harmonize the signature rejection handling for tags and commits.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+STDIN.each do |line|
+  rev_old, rev_new, ref = line.split
+
+  if not repo.config['hooks.allowunsignedtags']
```
The top comment states that the config options for this is `hooks.allowcommitsonmaster`.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+      # least two parents, and one of them has to be the old target.
+      parents = repo.lookup(rev_new).parent_ids
+      if parents.length < 2 or
+          not parents.include?(rev_old)
```
The initial commit on `master` also needs to be pushable.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+@collaborators = YAML.load_file(repo.path + 'collaborators.yaml')
+
+def is_allowed_signer(keyid)
+  keys = GPGME::Key.find(keyid)
```
`GPGME::Key.find` does not work that way. The first arg should be `:public` to find public keys, and the needle should be in the second arg.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+  walker = Rugged::Walker.new(repo)
+
+  # Get all new commits on branch ref, even if it's a new branch.
+  walker.push(rev_new)
```
When pushing a branch deletion, `rev_new` is zeroes, which cannot be pushed.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+
+    if commit.oid.to_i(16) == 0
+      # A commit with a hash full of zeros is a deletion of a ref.
+      commit_type = :delete
```
I don't think ref deletions can appear in the actual commit graph.

notCalle commented on git-hooks/enforce-workflow.rb:
```diff
@@ -0,0 +1,236 @@
+      puts "*** Unknown type of update to ref #{ref} of type #{commit_type}."
+    end
+
+    if commit_count == 0
```
This condition is always false unless moved outside the `walker.each` block.

notCalle reviewed (changes requested):
In general very difficult to read, because there is a `.each do ... end` spanning > 100 lines that does almost everything in deeply nested logic structures.

The overall design is insecure because it accepts by default. The default should be to reject, unless the specific circumstances where a push is allowed has been verified.

Probably needs a `Gemfile` to describe its dependencies.

creideiki committed:
Correct calling of GPGME::Key.find

creideiki committed:
Exit if signature verification raises an exception

creideiki committed:
Test correct configuration keys

Copy-paste-damage meant that the key hooks.allowunsignedtags was
tested in some inappropriate places. Replace those instances with the
correct key.

creideiki committed:
Remove leftover debug puts

creideiki committed:
Move deletion handling to top of loop

The commit graph walker doesn't handle deletions, so break out the
handling of those to before the walk.

creideiki committed:
Fix configuration reading

Unlike what I thought, Rugged doesn't parse the values in the
repository configuration, instead just returning them as strings.
This means that they have to be compared to the string 'true', and
not assumed to be booleans.

creideiki committed:
Move the empty walker case outside the walker

The end of the walker loop had a block of code intended to be run if
the walker didn't find any commits. However, since it was inside the
actual walker loop, it was never run. Move it outside.

creideiki committed:
Accept initial creation of master

Even if master is only supposed to accept merges of feature branches,
it needs to start somewhere, in a bare commit. If master doesn't exist
yet, let bare commits through to the signature check.

creideiki committed:
When hiding refs from the walker, only hide heads

Tags should point to commits reachable from heads anyway, and remotes
have a different API in Rugged and never worked.

creideiki committed:
Fix some easy RuboCop warnings

Change as suggested by the RuboCop tool for the following tests:

Style/AndOr
Style/HashSyntax
Style/IfInsideElse
Style/MultilineOperationIndentation
Style/NegatedIf
Style/Not
Style/NumericPredicate
Style/PredicateName
Style/RedundantReturn
Style/RegexpLiteral
Style/SpaceInsideBlockBraces
Style/StringLiterals

creideiki commented:
Pushed 9045ea1 which I think should solve all actual bugs before I start restructuring the logic. However, the call to `Rugged::Commit.extract_signature` likes to corrupt the last couple of bytes of the plaintext on my system, so it makes spurious claims of signatures being bad.

creideiki commented:
Signature validation fails due to a bug in libgit2: libgit2/libgit2#4118

creideiki committed:
Correct typo in error message for unauthorised signers

creideiki commented:
I no longer see random signature validation failures as of Rugged 0.26.0b5.

creideiki committed:
Move enforce-workflow hook to its own directory

To prepare for somewhat more sane Ruby packaging, put each hook in its
own subdirectory.

creideiki committed:
Specify Ruby gem dependencies in a Gemfile

Rugged <0.26.0b5 bundles a buggy libgit2. GPGME 2.0.12 is the only one
I've tested with, and it's over a year old at this point.

creideiki committed:
Output signer username when signature is good

creideiki committed:
Break out deletion handling to a separate function

Start refactoring the gigantic loop by breaking out independent tests
into their own functions. Begin with the check for deleting a ref.

Make the repository object global to let other functions access its
configuration.

creideiki committed:
Break out handling of commits to master to a separate function

Commits to master are subject to special checks, in case we only want
merges of feature branches to master and no raw commits. Move these
checks to a separate function.

creideiki committed:
Break out commit graph walking code in separate functions

The code for walking the commit graph, and for handling tags (which
are detected by the walk visiting zero objects), was a long spaghetti
mess. Break it out into separate functions for walking and handling
the two different types of tags.

creideiki committed:
Unify exit conditions

Make all exit conditions in the main loop be of the form
"exit ... unless accept...".

creideiki committed:
Use single-quoted strings when not interpolating

creideiki committed:
Order dependencies alphabetically, to make RubyCop happy

creideiki commented:
This should be ready for another look now.

notCalle commented:
Pushing the initial commit to an empty repo still fails.
```
remote: *** Accepting creation of master branch.
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

creideiki committed:
Actually allow creating master, not just saying that we do

creideiki commented:
I have no idea what you are talking about. It's always worked wonderfully, with no bugs whatsoever.

notCalle commented:
Using `git` >= 2.11 seems to break `pre-receive`, so that should be noted somewhere... 😞

`git` 2.11.0 and 2.12.2:

```
Counting objects: 3, done.
Writing objects: 100% (3/3), 845 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
remote: hooks/pre-receive:126:in `push': object not found - no match for id (f5f1e8e9f4511a7c83de6eefb0b99c9b5331e03a) (Rugged::OdbError)
remote: 	from hooks/pre-receive:126:in `walk_graph'
remote: 	from hooks/pre-receive:288:in `block in <main>'
remote: 	from hooks/pre-receive:266:in `each'
remote: 	from hooks/pre-receive:266:in `<main>'
remote: *** Accepting creation of master branch.
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

`git` 2.10.1:

```
Counting objects: 3, done.
Writing objects: 100% (3/3), 845 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
remote: *** Accepting creation of master branch.
remote: *** Good signature on commit f5f1e8e9f4511a7c83de6eefb0b99c9b5331e03a by calle@discord.bofh.se (1764CE72729D4C843AF6999A78D9BF4A13F75F59).
To /Users/calle/Workspace/signed-only.git
 * [new branch]      master -> master
```

notCalle commented:
Pushing a new (signed) tag fails:
```
Counting objects: 1, done.
Writing objects: 100% (1/1), 793 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
remote: *** Tag refs/tags/with-readme already exists.
remote: *** Modifying a tag is not allowed in this repository.
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] with-readme -> with-readme (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

creideiki committed:
Fix checking of tag signatures

The old code basically didn't work at all. Fixes:
* Send the ID of the new object the tag will point to, so we can check
  the signature.
* Invert test for when to complain about re-defining a tag.
* Use Rugged::Tag.extract_signature to extract the signature, instead
  of Rugged::Commit.extract_signature which doesn't work on tags.
* Fix variable scope in the signature check, to avoid reading from
  undefined variables.

notCalle commented:
Pushing a new unsigned tag fails with stack trace instead of proper error message.
```
Counting objects: 1, done.
Writing objects: 100% (1/1), 181 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
remote: /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/ctx.rb:447:in `verify': No data (GPGME::Error::NoData)
remote: 	from /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/crypto.rb:312:in `block in verify'
remote: 	from /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/ctx.rb:79:in `new'
remote: 	from /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/crypto.rb:311:in `verify'
remote: 	from hooks/pre-receive:237:in `allow_annotated_tag?'
remote: 	from hooks/pre-receive:308:in `block in <main>'
remote: 	from hooks/pre-receive:270:in `each'
remote: 	from hooks/pre-receive:270:in `<main>'
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] with-readme -> with-readme (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

creideiki committed:
Only try to validate tag signatures that actually exist

notCalle commented on git-hooks/enforce-workflow/enforce-workflow.rb:
```diff
@@ -0,0 +1,316 @@
+      puts "*** Good signature on tag #{ref} by #{signer} (#{fingerprint})."
+      return true
+    else
+      puts "*** Rejecting tag #{ref} due to lack of a valid GPG signature."
```
Lies! It has a valid signature, but the signer is not authorized (see `walk_graph`). 

Also, maybe consider a DRYer approach to signature validation and reporting?

creideiki committed:
Correct tag signature validation

Set default variable values to nil, not the empty string, since we'll
be using them in boolean expressions.

Use the correct error message for signed but unauthorised tags.

creideiki committed:
Break out signature checking into its own function

Required keyword arguments require Ruby 2.1, but we need to be able to
run on CentOS 7's Ruby 2.0. Instead, use a default value for `type`
and error out if it is used.

creideiki committed:
Re-introduce fix from 8f9bcf6 to split signature checking code

When the signature checking code was broken out into its own function,
the fix for unsigned objects in 8f9bcf6 was lost. Re-introduce it.
